src/libraries/Common/src/System/Threading/Tasks/TaskTimeoutExtensions.cs

```diff
@@ -0,0 +1,42 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
+
+namespace System.Threading.Tasks
+{
+    /// <summary>
+    /// Task timeout helper based on https://devblogs.microsoft.com/pfxteam/crafting-a-task-timeoutafter-method/
+    /// </summary>
+    internal static class TaskTimeoutExtensions
+    {
+        public static async Task WithCancellation(this Task task, CancellationToken cancellationToken)
+        {
+            if (task is null)
+            {
+                throw new ArgumentNullException(nameof(task));
+            }
+
+            if (task.IsCompleted || !cancellationToken.CanBeCanceled)
+            {
+                await task.ConfigureAwait(false);
+                return;
+            }
```

Comment:
These should be in a non-async method, and optimize the cases where the original task can just be returned, e.g.
```C#
public static Task WithCancellation(this Task task, CancellationToken cancellationToken)
{
    if (task is null)
    {
        throw new ArgumentNullException(nameof(task));
    }

    if (task.IsCompleted || !cancellationToken.CanBeCanceled)
    {
        return task;
    }

    return WaitAsync(task, cancellationToken);

    static async Task WaitAsync(this Task task, CancellationToken cancellationToken)
    {
        ...
    }
}
````

Reply:
Given that `Task.FromCanceled` manifests as TaskCanceledException, perhaps it might make sense to also throw the same exception type in the asynchronous part for consistency?

Reply:
I'm ok with either. I really don't want devs relying on this being a TaskCanceledException rather than OperationCanceledexception, but obviously some may.

I wish we never shipped TaskCanceledException.

Reply:
I ended up doing this

```csharp
throw new TaskCanceledException(Task.FromCanceled(cancellationToken));
```